File: 081_search_in_rotated_sorted_array_ii/Solution.java
class Solution {
    public boolean search(int[] nums, int target) {
        int arrLen = nums.length;

        int l = 0, r = arrLen - 1;
        while (l <= r) {
            while (l < r && nums[l] == nums[l + 1]) l++;
            while (l < r && nums[r] == nums[r - 1]) r--;
            int mid = (l + r) / 2;
            if (nums[mid] == target) return true;
            // nums[l] <= nums[mid] nums[l]等于nums[mid]且有序的情况必在左侧，
            // 否则将违反只旋转一次的规定
            if (nums[mid] < nums[r]) {
                if (nums[mid] < target && target <= nums[r]) l = mid + 1;
                else r = mid - 1;
            } else {
                if (nums[l] <= target && target < nums[mid]) r = mid - 1;
                else l = mid + 1;
            }
        }

        return false;
    }

    public static void main(String[] args) {
        int[] nums = {2, 5, 6, 0, 0, 1, 2};
        System.out.println(new Solution().search(nums, 2));
    }
}

File: 081_search_in_rotated_sorted_array_ii/Solution.py
class Solution:
    def search(self, nums: list, target: int) -> bool:
        # return target in nums
        arr_len = len(nums)
        l, r = 0, arr_len - 1

        while l <= r:
            while l < r and nums[l] == nums[l + 1]: l += 1
            while l < r and nums[r] == nums[r - 1]: r -= 1
            mid = (l + r) // 2
            if nums[mid] == target: return True
            if nums[mid] < nums[r]:
                if nums[mid] < target <= nums[r]: l = mid + 1
                else: r = mid - 1
            else:
                if nums[l] <= target < nums[mid]: r = mid - 1
                else: l = mid + 1

        return False


if __name__ == '__main__':
    nums = [4, 5, 6, 7, 0, 1, 2]
    print(Solution().search(nums, 2))

